onnxruntime/python/tools/kernel_explorer/kernels/rocm/gemm_softmax_gemm_permute.cu

```diff
@@ -0,0 +1,288 @@
+#include "python/tools/kernel_explorer/device_array.h"
+#include "python/tools/kernel_explorer/kernel_explorer_interface.h"
+
+#include <vector>
+
+namespace py = pybind11;
+
+using namespace onnxruntime::contrib::rocm;
+
+namespace onnxruntime {
+
+template <typename T>
+class IGemmSoftmaxGemmPermuteKernelExplorer : public IKernelExplorer {
+ public:
+  IGemmSoftmaxGemmPermuteKernelExplorer(
+      int64_t batch,
+      int64_t seqlen,
+      int64_t total_seqlen,
+      int64_t num_heads,
+      int64_t head_size,
+      int64_t mask_dim,
+      double scale,
+      DeviceArray& Q,
+      DeviceArray& K,
+      DeviceArray& V,
+      std::optional<DeviceArray>& attn_bias,
+      std::optional<DeviceArray>& attn_mask,
+      DeviceArray& out) {
+    ROCBLAS_CALL_THROW(rocblas_create_handle(&rocblas_handle_));
+
+    attn_.batch_size = batch;
+    attn_.sequence_length = seqlen;
+    attn_.kv_sequence_length = seqlen;        // NOTE: not used
+    attn_.past_sequence_length = 0;           // NOTE: not used
+    attn_.original_past_sequence_length = 0;  // NOTE: not used
+    attn_.total_sequence_length = total_seqlen;
+    attn_.max_sequence_length = -1;  // TODO: set
+    attn_.hidden_size = num_heads * head_size;
+    attn_.head_size = head_size;
+    attn_.v_hidden_size = attn_.hidden_size;  // Q,K,V hidden size must agree now
+    attn_.v_head_size = attn_.head_size;      // Q,K,V hidden size must agree now
+    attn_.num_heads = num_heads;
+    attn_.is_unidirectional = false;
+    attn_.past_present_share_buffer = false;
+    attn_.do_rotary = false;
+    attn_.mask_filter_value = -10000.0f;
+    attn_.scale = scale;
+    if (mask_dim == 0) {
+      attn_.mask_type = contrib::MASK_NONE;
+    } else if (mask_dim == 2) {
+      attn_.mask_type = contrib::MASK_2D_KEY_PADDING;
+    } else if (mask_dim == 3) {
+      attn_.mask_type = contrib::MASK_3D_ATTENTION;
+    } else if (mask_dim == 4) {
+      attn_.mask_type = contrib::MASK_4D_MEGATRON;
+    } else {
+      ORT_ENFORCE(false, "mask type not supported");
+    }
+
+    device_prop = GetEp()->GetDeviceProp();
+
+    params_.tuning_ctx = TuningContext();
+    params_.stream = Stream();
+    params_.handle = rocblas_handle_;
+    params_.attention = &attn_;
+    params_.device_prop = &device_prop;
+    params_.scale = scale;
+
+    params_.q_buffer = reinterpret_cast<T*>(Q.ptr());
+    params_.k_buffer = reinterpret_cast<T*>(K.ptr());
+    params_.v_buffer = reinterpret_cast<T*>(V.ptr());
+    if (attn_bias.has_value()) {
+      params_.bias_buffer = reinterpret_cast<T*>(attn_bias->ptr());
+    }
+    if (attn_mask.has_value()) {
+      params_.mask_index_buffer = reinterpret_cast<int*>(attn_mask->ptr());
+      if (mask_dim == 2) {
+        params_.mask_index_dims = {batch, total_seqlen};
+      } else if (mask_dim == 3) {
+        params_.mask_index_dims = {batch, seqlen, total_seqlen};
+      } else if (mask_dim == 4) {
+        int max_seqlen = 1024;  // TODO:
```

Comment:
Add a comment that kernel explorer only tests max_seq_len==1024 for 4D mask. I think the kernel does not have such limitation.